client/keystore/src/lib.rs

```diff
@@ -276,11 +277,87 @@ impl Store {
 		buf.push(key_type + key.as_str());
 		Some(buf)
 	}
+
+	/// Signs a message with the private key that matches
+	/// the public key passed.
+	fn _sign_with(
+		&self,
+		id: KeyTypeId,
+		key: &CryptoTypePublicPair,
+		msg: &[u8],
+	) -> std::result::Result<Vec<u8>, String> {
+		match key.0 {
+			ed25519::ED25519_CRYPTO_ID => {
+				let pub_key = ed25519::Public::from_slice(key.1.as_slice());
+				let key_pair: ed25519::Pair = self
+					.key_pair_by_type::<ed25519::Pair>(&pub_key, id)
+					.map(Into::into)
+					.map_err(|e| e.to_string())?;
+				return Ok(<[u8; 64]>::from(key_pair.sign(msg)).to_vec());
+			}
+			sr25519::SR25519_CRYPTO_ID => {
+				let pub_key = sr25519::Public::from_slice(key.1.as_slice());
+				let key_pair: sr25519::Pair = self
+					.key_pair_by_type::<sr25519::Pair>(&pub_key, id)
+					.map(Into::into)
+					.map_err(|e| e.to_string())?;
+				return Ok(<[u8; 64]>::from(key_pair.sign(msg)).to_vec());
+			}
+			_ => Err(String::from("Key kind invalid"))
+		}
+	}
 }
 
 impl BareCryptoStore for Store {
+	fn get_supported_keys(&self, id: KeyTypeId, keys: Vec<CryptoTypePublicPair>) -> std::result::Result<Vec<CryptoTypePublicPair>, String> {
+		let ed25519_existing_keys: Vec<Vec<u8>> = self.public_keys_by_type::<ed25519::Public>(id)
+								.map(|keys| keys.iter().map(|k| k.to_raw_vec()).collect())
+								.map_err(|e| e.to_string())?;
+		let sr25519_existing_keys: Vec<Vec<u8>> = self.public_keys_by_type::<sr25519::Public>(id)
+								.map(|keys| keys.iter().map(|k| k.to_raw_vec()).collect())
+								.map_err(|e| e.to_string())?;
+
+		Ok(keys.iter().filter_map(|k| {
+			match k.0 {
+				sr25519::SR25519_CRYPTO_ID if sr25519_existing_keys.contains(&k.1.to_vec()) => Some(k),
+				ed25519::ED25519_CRYPTO_ID if ed25519_existing_keys.contains(&k.1.to_vec()) => Some(k),
+				_ => None
+			}
+		}).cloned().collect::<Vec<_>>())
+	}
+
+	fn get_keys(&self, id: KeyTypeId) -> std::result::Result<Vec<CryptoTypePublicPair>, String> {
+		let ed25519_existing_keys: Vec<Vec<u8>> = self.public_keys_by_type::<ed25519::Public>(id)
+								.map(|keys| keys.iter().map(|k| k.to_raw_vec()).collect())
+								.map_err(|e| e.to_string())?;
+		let sr25519_existing_keys: Vec<Vec<u8>> = self.public_keys_by_type::<sr25519::Public>(id)
+								.map(|keys| keys.iter().map(|k| k.to_raw_vec()).collect())
+								.map_err(|e| e.to_string())?;
+
+		let mut keys: Vec<CryptoTypePublicPair> = vec![];
+		keys.extend(sr25519_existing_keys.iter()
+					.cloned()
+    				.map(|k| (sr25519::SR25519_CRYPTO_ID, k)));
+		keys.extend(ed25519_existing_keys.iter()
+    				.cloned()
+					.map(|k| (ed25519::ED25519_CRYPTO_ID, k))
+					.collect::<Vec<_>>());
+		Ok(keys)
+
```

Comment:
Maybe use `.chain` instead? And do that internal mapping already in the first iteration? Would save you a few vec allocations..


```rust

let ed_keys = self.public_keys_by_type::<ed25519::Public>(id)
	.map_err(|e| e.to_string())?
	.iter().map(|k| (ed25519::ED25519_CRYPTO_ID, k.to_raw_vec()));

let sr_keys = self.public_keys_by_type::<sr25519::Public>(id)
	.map_err(|e| e.to_string())?
	.iter().map(|k| (sr25519::SR25519_CRYPTO_ID, k.to_raw_vec()));

Ok(ed_keys.chain(sr_keys).collect::Vec<_>());
```

Wouldn't that do it?

Reply:
Done,,, also had to use `cloned` here